lib/sdk.js

```diff
@@ -0,0 +1,414 @@
+  return fs.readdirAsync(path.join(paths.sdk, getPlatform()))
+    .then((values) => new Promise((resolve, reject) => {
+      for (var i = 0; i < values.length; i++) {
+        if (values[i].match(/^toolchain\-/)) {
+          return resolve(path.join(paths.sdk, getPlatform(), values[i]));
+        }
+      }
+      return reject(new Error('No toolchain found.'));
+    }));
+};
+
+// Checks is CHECKSUM file in our SDK equals our expected checksum.
+// This will resolve with checking that the SDK exists and matches the checksum.
+exports.checkSdk = (checksumVerify) => {
+  var dir = path.join(paths.sdk, getPlatform());
+  return fs.readFileAsync(path.join(dir, 'CHECKSUM'), 'utf-8')
+    .then(checksum => ({
+      exists: true,
+      checked: checksumVerify === checksum,
+      path: dir,
+    }), () => ({
+      exists: false,
+      checked: false,
+      path: dir,
+    }));
+};
+
+exports.checkRustlib = (rustv, checksumVerify) => {
+  var dir = path.join(paths.rustlib, rustv);
+  return fs.readFileAsync(path.join(dir, 'CHECKSUM'), 'utf-8')
+    .then(checksum => ({
+      exists: true,
+      checked: checksumVerify === checksum,
+      path: dir,
+    }), () => ({
+      exists: false,
+      checked: false,
+      path: dir,
+    }));
+};
+
+exports.installSdk = () => {
+  var url = sdkPath[getPlatform()];
+  var checksumVerify = null;
+
+  return downloadString(url + '.sha256')
+    .then((checksum) => {
+      checksumVerify = checksum;
+      return exports.checkSdk(checksumVerify);
+    })
+    .then((check) => {
+      if (check.exists && check.checked) {
+        log.info('Latest SDK already installed.');
+        return;
+      } else if (!check.exists) {
+        log.info('Installing SDK...');
+      } else {
+        log.info('Updating SDK...');
+      }
+
+      return fs.mkdirpAsync(path.join(osenv.home(), '.tessel/sdk'))
+        .then(() => extractSdk(checksumVerify, path.basename(url), download(url)));
+    });
+};
+
+exports.installRustlib = () => {
+  var url = null;
+  var checksumVerify = null;
+  var rustv = null;
+  var pkgname = null;
+
+  return exports.rustVersion()
+    .then(_rustv => {
+      rustv = _rustv;
+      pkgname = 'MIPS libstd v' + rustv;
+      url = rustlibUrl.replace('VERSION', rustv);
+
+      return downloadString(url + '.sha256');
+    })
+    .catch(() => {
+      throw new Error(`Could not find a MIPS libstd matching your current Rust version (${rustv}). Only stable Rust versions >= 1.11.0 are supported.`);
+    })
+    .then(checksum => {
+      checksumVerify = checksum;
+      return exports.checkRustlib(rustv, checksumVerify);
+    })
+    .then(check => {
+      if (check.exists && check.checked) {
+        log.info(`Latest ${pkgname} already installed.`);
+        return;
+      } else if (!check.exists) {
+        log.info(`Installing ${pkgname}...`);
+      } else {
+        log.info(`Updating ${pkgname}...`);
+      }
+
+      return fs.mkdirpAsync(paths.rustlib)
```

Comment:
Eventually t2-cli can be refactored to use fs-extra-promise, but until it's done across the entire project, all fs operations should use fs-extra. (This PR is not an appropriate vehicle for such changes)